### ISSUE TITLE
Reset fails on file to folder transition

### ISSUE THREAD

**Issue by vmg** [libgit2/libgit2#3541]
Pretty easy to reproduce. Have a repo that looks like this:

```
100644 blob e69de29bb2d1d6434b8b29ae775ad8c2e48c5391    README
100644 blob e69de29bb2d1d6434b8b29ae775ad8c2e48c5391    folder
```

Then remove the file called `folder` and replace it with an actual folder.

```
100644 blob e69de29bb2d1d6434b8b29ae775ad8c2e48c5391    README
100644 blob 03ed41df1178c190a425bcbc5ba9363acbd5533b    folder/FILE
```

Calling `git_reset(HARD)` through libgit2 will fail with `folder/FILE appears as both a file and a directory`.

How do we work around this? We're calling `git_read_tree` into the existing index after the checkout, without clearing the existing index. Is this intended? `read_tree` eventually clears the index, but _after_ the entries have been loaded into a temporary vector.

cc @ethomson @carlosmn @ptoomey3

**Comment by vmg** [comment]
Actually, `git_read_tree` is a red herring. The error happens earlier during the checkout:

```
#0  index_insert (index=index@entry=0x555556036ac0, entry_ptr=entry_ptr@entry=0x7fffffffcd70, trust_path=trust_path@entry=true, 
    trust_mode=trust_mode@entry=true, replace=1) at /home/vmg/src/rugged/vendor/libgit2/src/index.c:1316
#1  0x00007ffff6145e36 in git_index_add (index=0x555556036ac0, source_entry=source_entry@entry=0x7fffffffcda0)
    at /home/vmg/src/rugged/vendor/libgit2/src/index.c:1558
#2  0x00007ffff6113474 in checkout_update_index (file=0x55555603e4f0, st=<optimized out>, data=<optimized out>)
    at /home/vmg/src/rugged/vendor/libgit2/src/checkout.c:1576
#3  0x00007ffff6115e90 in checkout_blob (file=<optimized out>, data=<optimized out>)
    at /home/vmg/src/rugged/vendor/libgit2/src/checkout.c:1738
#4  checkout_create_the_new (data=<optimized out>, actions=<optimized out>)
    at /home/vmg/src/rugged/vendor/libgit2/src/checkout.c:1842
#5  git_checkout_iterator (target=0x1, index=0x55555603d0e0, opts=0x7fffffffd338)
    at /home/vmg/src/rugged/vendor/libgit2/src/checkout.c:2566
#6  0x00007ffff6116d12 in git_checkout_tree (repo=<optimized out>, repo@entry=0x555556030fe0, treeish=<optimized out>, 
    opts=opts@entry=0x7fffffffd500) at /home/vmg/src/rugged/vendor/libgit2/src/checkout.c:2697
#7  0x00007ffff6155a70 in reset (repo=repo@entry=0x555556030fe0, target=target@entry=0x555556036840, 
    to=0x555556031170 "69cb3ccd1af462c5c9554477ed697d863b1ac7f6", reset_type=reset_type@entry=GIT_RESET_HARD, 
    checkout_opts=checkout_opts@entry=0x0) at /home/vmg/src/rugged/vendor/libgit2/src/reset.c:157
#8  0x00007ffff6155d54 in git_reset (repo=repo@entry=0x555556030fe0, target=target@entry=0x555556036840, 
    reset_type=reset_type@entry=GIT_RESET_HARD, checkout_opts=checkout_opts@entry=0x0)
    at /home/vmg/src/rugged/vendor/libgit2/src/reset.c:189
```

**Comment by carlosmn** [comment]
It looks like we might be doing the checkout in the wrong order. We first move HEAD and then perform the checkout, but usually what you should do is perform the checkout first, and then update the rest of the structures. This serves partly as error-recovery, but it should also help here to realise that there are no changes between the workdir and HEAD's tree.

**Comment by carlosmn** [comment]
Fixed via #3542
